citation-categories/java/src/main/java/org/linqs/psl/examples/citationcategories/Run.java

```diff
@@ -0,0 +1,246 @@
+
+        // Add one rule for each category.
+        for (int cat = 1; cat <= NUM_CATEGORIES; cat++) {
+            String stringRule = String.format("1.0: HasCat(A, '%d') & Link(A, B) >> HasCat(B, '%d') ^2", cat, cat);
+            model.addRule(stringRule);
+        }
+
+        model.addRule("HasCat(A, +C) = 1 .");
+
+        model.addRule("0.001: !HasCat(A, N) ^2");
+
+        log.debug("model: {}", model);
+    }
+
+    /**
+     * Load data from text files into the DataStore.
+     * Three partitions are defined and populated: observations, targets, and truth.
+     * Observations contains evidence that we treat as background knowledge and use to condition our inferences.
+     * Targets contains the inference targets - the unknown variables we wish to infer.
+     * Truth contains the true values of the inference variables and will be used to evaluate the model's performance.
+     */
+    private void loadData() {
+        log.info("Loading data into database");
+        String[] wList = new String[]{"learn", "eval"}; 
+        for (String type : wList ) {
+            Partition obsPartition = dataStore.getPartition(type + "_observations");
+            Partition targetsPartition = dataStore.getPartition(type + "_targets");
+            Partition truthPartition = dataStore.getPartition(type + "_truth");
+
+            Inserter inserter = dataStore.getInserter(model.getStandardPredicate("Link"), obsPartition);
+            inserter.loadDelimitedDataTruth(Paths.get(DATA_PATH, type, "link_obs.txt").toString());
+
+            inserter = dataStore.getInserter(model.getStandardPredicate("HasCat"), obsPartition);
+            inserter.loadDelimitedDataTruth(Paths.get(DATA_PATH, type, "hasCat_obs.txt").toString());
+
+            inserter = dataStore.getInserter(model.getStandardPredicate("HasCat"), targetsPartition);
+            inserter.loadDelimitedData(Paths.get(DATA_PATH, type, "hasCat_targets.txt").toString());
+
+            inserter = dataStore.getInserter(model.getStandardPredicate("HasCat"), truthPartition);
+            inserter.loadDelimitedDataTruth(Paths.get(DATA_PATH, type, "hasCat_truth.txt").toString());
+        }
+    }
+
+    /**
+     * Use the training data to learn weights for our rules and store them back in the model object.
+     */
+    private void learnWeights() {
+        log.info("Starting weight learning");
+
+        Partition obsPartition = dataStore.getPartition(PARTITION_LEARN_OBSERVATIONS);
+        Partition targetsPartition = dataStore.getPartition(PARTITION_LEARN_TARGETS);
+        Partition truthPartition = dataStore.getPartition(PARTITION_LEARN_TRUTH);
+
+        // This database contains all the ground atoms (targets) that we want to infer.
+        // It also includes the observed data (because we will run inference over this db).
+        Set<StandardPredicate>set = new HashSet<StandardPredicate>();
+        set.add(model.getStandardPredicate("Link"));
```

Comment:
Instead of adding all the closed predicates (which may be a lot for ones like KGI), how about you get all the predicates from datastore and then remove the open ones (which is usually much less).
Ex:
```
        // Get all the predicates from the data store, and remove the open ones.                                                        
        Set<StandardPredicate> closedPredicates = dataStore.getRegisteredPredicates();                                                 
        closedPredicates.remove(model.getStandardPredicate("Cat"));                                                                    
        closedPredicates.remove(model.getStandardPredicate("Rel"));
```

Reply:
I did not implement it this example because number of closed predicates were very less. I have implemented it in the KGI.